agent/consul/state/config_entry.go

```diff
@@ -371,6 +373,100 @@ var serviceGraphKinds = []string{
+func (s *Store) discoveryChainTargets(ws memdb.WatchSet, dc, service string, entMeta *structs.EnterpriseMeta) (uint64, []structs.ServiceName, error) {
+	source := structs.NewServiceName(service, entMeta)
+	req := discoverychain.CompileRequest{
+		ServiceName:         source.Name,
+		EvaluateInNamespace: source.NamespaceOrDefault(),
+
+		// TODO(freddy) : Should these be anything other than the known DC?
+		EvaluateInDatacenter: dc,
+		UseInDatacenter:      dc,
+	}
+	idx, chain, err := s.ServiceDiscoveryChain(ws, source.Name, entMeta, req)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to fetch discovery chain for %q: %v", source.String(), err)
+	}
+
+	var resp []structs.ServiceName
+	for _, t := range chain.Targets {
+		em := structs.EnterpriseMetaInitializer(t.Namespace)
+		target := structs.NewServiceName(t.Service, &em)
+
+		// TODO (freddy): Allow upstream DC and encode in response
+		if t.Datacenter == dc {
+			resp = append(resp, target)
+		}
+	}
+	return idx, resp, nil
+}
+
+// discoveryChainSources will return a list of services whose discovery chains have the input service as a target
+func (s *Store) discoveryChainSources(ws memdb.WatchSet, tx ReadTxn, dc string, destination structs.ServiceName) (uint64, []structs.ServiceName, error) {
+	queue := []structs.ServiceName{destination}
+
+	seenLink := make(map[structs.ServiceName]bool)
+	for len(queue) > 0 {
+		// The "link" index returns config entries that reference a service
+		iter, err := tx.Get(configTableName, "link", queue[0].ToServiceID())
+		if err != nil {
+			return 0, nil, err
+		}
+		ws.Add(iter.WatchCh())
+
+		for raw := iter.Next(); raw != nil; raw = iter.Next() {
+			entry := raw.(structs.ConfigEntry)
+
+			sn := structs.NewServiceName(entry.GetName(), entry.GetEnterpriseMeta())
+			if !seenLink[sn] {
+				seenLink[sn] = true
+				queue = append(queue, sn)
+			}
+		}
+		queue = queue[1:]
+	}
+
+	var (
+		maxIdx uint64
+		resp   []structs.ServiceName
+	)
+
+	// Only return the services that directly target the destination
+	seenSource := make(map[structs.ServiceName]bool)
+	for sn, _ := range seenLink {
+		req := discoverychain.CompileRequest{
+			ServiceName:         sn.Name,
+			EvaluateInNamespace: sn.NamespaceOrDefault(),
+
+			// TODO(freddy) : Should these be anything other than the known DC?
+			EvaluateInDatacenter: dc,
+			UseInDatacenter:      dc,
+		}
+		idx, chain, err := s.ServiceDiscoveryChain(ws, sn.Name, &sn.EnterpriseMeta, req)
+		if err != nil {
+			return 0, nil, fmt.Errorf("failed to fetch discovery chain for %q: %v", sn.String(), err)
+		}
+
+		for _, t := range chain.Targets {
+			em := structs.EnterpriseMetaInitializer(t.Namespace)
+			candidate := structs.NewServiceName(t.Service, &em)
+
+			if !candidate.Matches(&destination) {
+				continue
+			}
+			if idx > maxIdx {
+				maxIdx = idx
+			}
+			if !seenSource[sn] {
+				seenSource[sn] = true
+				resp = append(resp, sn)
+			}
+		}
+	}
+	return maxIdx, resp, nil
```

Comment:
Should this response be deterministically sorted at all?

Reply:
I don't really see a need for that. We sort final output in the HTTP endpoint.

Or do you mean for the test assertions?

Reply:
Sometimes it helps for either. The random map iteration dumping results into a slice can be surprising when it bubbles out to an API endpoint.

Since this is presumably entirely an internal API however it's not so bad.